src/main/java/com/conveyal/datatools/manager/models/transform/PreserveCustomFieldsTransformation.java

```diff
@@ -0,0 +1,132 @@
+            String hashKey = StringUtils.join(customCsvKeyValues, "_");
+            lookup.put(hashKey, nextLine);
+        }
+        return lookup;
+    }
+
+    @Override
+    public void transform(FeedTransformZipTarget zipTarget, MonitorableJob.Status status) {
+        String tableName = table + ".txt";
+        Path targetZipPath = Paths.get(zipTarget.gtfsFile.getAbsolutePath());
+
+        // TODO: is there a better way to do this than using a Singleton collector?
+        specTable = Arrays.stream(Table.tablesInOrder)
+                .filter(t -> t.name.equals(table))
+                .collect(toSingleton());
+        tablePrimaryKeys = specTable.getPrimaryKeyNames();
+
+        try( FileSystem targetZipFs = FileSystems.newFileSystem(targetZipPath, (ClassLoader) null) ){
+            List<String> specTableFields = specTable.specFields().stream().map(f -> f.name).collect(Collectors.toList());
+            Path targetTxtFilePath = getTablePathInZip(tableName, targetZipFs);
+
+            // TODO: There must be a better way to do this.
+            InputStream is = Files.newInputStream(targetTxtFilePath);
+            final File tempFile = File.createTempFile(tableName + "-temp", ".txt");
+            File output = File.createTempFile(tableName + "-output-temp", ".txt");
+            FileOutputStream out = new FileOutputStream(tempFile);
+            IOUtils.copy(is, out);
+
+            CsvMapReader newEditorFileReader = new CsvMapReader(new FileReader(tempFile), CsvPreference.STANDARD_PREFERENCE);
+            CsvMapReader newCustomFileReader = new CsvMapReader(new StringReader(csvData), CsvPreference.STANDARD_PREFERENCE);
+
+            String[] customHeaders = newCustomFileReader.getHeader(true);
+            final String[] editorHeaders = newEditorFileReader.getHeader(true);
+
+            List<String> customFields = Arrays.stream(customHeaders).filter(h -> !specTableFields.contains(h)).collect(Collectors.toList());
+            if (customFields.size() == 0) return;
+            String[] fullHeaders = ArrayUtils.addAll(editorHeaders, customFields.toArray(new String[0]));
+
+            HashMap<String, Map<String, String>> lookup = createCsvHashMap(newCustomFileReader, customHeaders);
+            CsvMapWriter writer = new CsvMapWriter(new FileWriter(output), CsvPreference.STANDARD_PREFERENCE);
+            writer.writeHeader(fullHeaders);
+
+            Map<String, String> row;
+            while ((row = newEditorFileReader.read(editorHeaders)) != null) {
+                List<String> editorCsvPrimaryKeyValues = tablePrimaryKeys.stream()
+                    .map(row::get)
+                    .collect(Collectors.toList());
+
+                String hashKey = StringUtils.join(editorCsvPrimaryKeyValues, "_");
+                Map<String, String> customCsvValues = lookup.get(hashKey);
+                Map<String, String> finalRow = row;
```

Comment:
Only because variables in lambda expressions have to be final or effectively final. Is there a better way around this I'm missing? 
<img width="684" alt="image" src="https://github.com/ibi-group/datatools-server/assets/63798641/b096abfe-53bb-4615-955f-3a5c6346f131">